### PR TITLE
fix(git-refs): search hostRules by http url, not ssh.

### LINKED ISSUE #17837: hostRules matchHost doesnt match git ssh urls
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.171.0 

### If you're self-hosting Renovate, select which platform you are using.

GitHub Enterprise Server

### If you're self-hosting Renovate, tell us what version of the platform you run.

3.3.9

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

I've been trying to get [batect bundle](https://docs.renovatebot.com/modules/manager/batect/) support working for some time, but its never picked up new tags in the bundle repository. after a lot of deep-diving through the source code, it appears that this is because [the hostRules matcher](https://github.com/renovatebot/renovate/blob/main/lib/util/host-rules.ts#L95) tries to turn the given url from the batect manager into a `new URL`, which fails for ssh urls, and [the git-refs datasource](https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/git-refs/base.ts#L33) doesn't try to turn what may be a ssh git url into a http(s) one. This results in no hostRule being resolved, and the datasource being unable to list tags because it is unauthorized without a token, when a git-tags datasource points at a ssh url rather than an https one

It appears to me that the fix is to apply the same code as is used [here](https://github.com/renovatebot/renovate/blob/3b576edc25caf8edb461e22da94908fa8bc4ee3f/lib/util/git/url.ts#L21) to the url (/`packageSource`) passed into the git-tags datasource before trying to get `hostRules` for it

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Datasource unknown error (repository=authorisation/permissions.service)
       "datasource": "git-tags",
       "packageName": "git@github.dev.xero.com:authorisation/authorisation.shared.git",
       "err": {
         "task": {
           "commands": [
             "ls-remote",
             "git@github.dev.xero.com:authorisation/authorisation.shared.git"
           ],
           "format": "utf-8",
           "parser": "[function]"
         },
         "message": "Host key verification failed.\r\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\n",
         "stack": "Error: Host key verification failed.\r\nfatal: Could not read from remote repository.\n\nPlease make sure you have the correct access rights\nand the repository exists.\n\n    at Object.action (/usr/src/app/node_modules/simple-git/src/lib/plugins/error-detection.plugin.ts:42:29)\n    at PluginStore.exec (/usr/src/app/node_modules/simple-git/src/lib/plugins/plugin-store.ts:29:29)\n    at /usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:127:42\n    at new Promise (<anonymous>)\n    at GitExecutorChain.handleTaskData (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:124:14)\n    at GitExecutorChain.<anonymous> (/usr/src/app/node_modules/simple-git/src/lib/runners/git-executor-chain.ts:100:40)\n    at Generator.next (<anonymous>)\n    at fulfilled (/usr/src/app/node_modules/simple-git/dist/cjs/index.js:55:24)"
       }
DEBUG: Failed to look up dependency git@github.dev.xero.com:authorisation/authorisation.shared.git (repository=authorisation/permissions.service, packageFile=batect.yml, dependency=git@github.dev.xero.com:authorisation/authorisation.shared.git)```

</details>


### Have you created a minimal reproduction repository?

No reproduction repository

### PR DESCRIPTION
**Pull request by flo-everett-xero**
also, fix token url encoding

## Changes

resolve https://github.com/renovatebot/renovate/issues/17837. 
- ensure that git urls passed into `git-refs` `getRemoteUrlWithToken` is converted to https before searching against `hostRules`
- preserve the first `:` in a `hostRules` `token` so that it is treated as a `username:password` pair rather than just a `username` 

## Context

Closes #17837

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/util/git/url.ts [hunk 1]
```diff
@@ -29,12 +29,23 @@ export function getHttpUrl(url: string, token?: string): string {
...
 export function getRemoteUrlWithToken(url: string, hostType?: string): string {
-  const hostRule = hostRules.find({ url, hostType });
+  const httpUrl = getHttpUrl(url)
+  const hostRule = hostRules.find({ url: httpUrl, hostType });
 
   if (hostRule?.token) {
     logger.debug(`Found hostRules token for url ${url}`);
-
-    return getHttpUrl(url, encodeURIComponent(hostRule.token));
+    const colonIndex = hostRule.token.indexOf(":")
```
Comment (viceice) [inline lib/util/git/url.ts:37]: wrong, you should store as username and password instead. the this workaround isn't required.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/util/git/url.ts (hunk 1 above)
Comment (flo-everett-xero) [reply lib/util/git/url.ts:37]: fortunately, github accepts `username: x-access-token, password: gh_123456789` as basic auth credentials for the API, so using username/password would technically work. however, I do think its worth looking into the issue with the token

Comment (rarkins) [reply lib/util/git/url.ts:37]: If username@host.name is valid, it's strange if simple-git or maybe git itself force a check for password.

But if that is as it is then you propose prepending `x-access-token:`? And is that for every type of token, both PAT and app tokens?

Comment (flo-everett-xero) [reply lib/util/git/url.ts]: I would like to check though - will a `hostRule` like 
```
{
    "matchHost": "github.com",
    "token": "ghs_abcd1234",
}
```
work on its own for release notes retrieval? we're populating these values using a script. I want to make sure that this is a valid token for release notes, and doesn't need to be of the form
```
{
    "matchHost": "github.com",
    "token": "x-access-token:ghs_abcd1234",
}
```

Comment (flo-everett-xero) [reply lib/util/git/url.ts]: In response to the above, from testing, it does appear to work fine. I would like to request re-review of this PR now :)

**Review by viceice — COMMENTED** [review]
needs tests

**Comment by flo-everett-xero** [comment]
bumping for test re-run

File: lib/util/git/url.spec.ts [hunk 2]
```diff
@@ -144,5 +174,14 @@ describe('util/git/url', () => {
         'https://gitlab-ci-token:token@gitlab.com/some/repo.git'
       );
     });
+
+    it('returns https url with encoded github token', () => {
+      hostRules.find.mockReturnValueOnce({
+        token: 'token',
+      });
+      expect(getRemoteUrlWithToken('ssh://github.com/some/repo.git')).toBe(
+        'https://x-access-token:token@github.com/some/repo.git'
```
Comment (viceice) [inline lib/util/git/url.spec.ts:183]: is this new behavior? or does it return https url before too?

if it changed, i consider this a breaking, as some people use ssh with git-refs datasource

Comment (flo-everett-xero) [reply lib/util/git/url.spec.ts:183]: this is not new behaviour - you can see in the tests on line 143 and line 173 of this file that this was already the case

**Review by viceice — APPROVED** [review]
btw: if you use a token, the username doesn't matter for most git servers

https://github.com/renovatebot/renovate/blob/fb09a30b5e03e1e1211f2f7e3dc9cbc3cb8cb337/lib/util/git/auth.ts#L102

https://coolaj86.com/articles/vanilla-devops-git-credentials-cheatsheet/

**Comment by viceice** [comment]
we should probably do same here

https://github.com/renovatebot/renovate/blob/fb09a30b5e03e1e1211f2f7e3dc9cbc3cb8cb337/lib/util/git/auth.ts#L78-L80

**Comment by flo-everett-xero** [comment]
@viceice would you like me to make the changes in your comment as part of this PR?

**Pull request merged**